src/testing_util/openssl_assertions.h

```diff
@@ -0,0 +1,138 @@
+      "           Actual: %s", message, expected, actual);
+}
+
+// Returns `testing::AssertionSuccess` if `openssl_return` is true. Otherwise,
+// returns `testing::AssertionFailure` with a formatted message containing the
+// first OpenSSL error in the OpenSSL error queue.
+//
+// Defined as a predicate instead of a matcher since matchers must be purely
+// functional, but `ERR_get_error_line_data` has OpenSSL-internal side-effects.
+testing::AssertionResult IsOpenSslSuccess(const char* expr,
+                                          bool openssl_return) {
+  if (openssl_return) return testing::AssertionSuccess();
+
+  const char *file, *data;
+  int line, flags;
+  auto err = ERR_get_error_line_data(&file, &line, &data, &flags);
+
+  // If err == 0, then no error was available in the OpenSSL error queue.
+  if (err) {
+    auto message = absl::StrFormat("%s unexpectedly returned false and "
+                                   "produced OpenSSL error...", expr);
+    auto actual = absl::StrFormat("false, with error in %s:%d [%s]", file, line,
+                                  data);
+    return MakeFormattedAssertionFailure(message, "success", actual);
+  } else {
+    auto message = absl::StrFormat("%s unexpectedly returned false...", expr);
+    return MakeFormattedAssertionFailure(message, "success",
+                                         "false (no OpenSSL error)");
+  }
+}
+
+// Returns `testing::AssertionSuccess` if `openssl_return` is false and the
+// first OpenSSL error in the OpenSSL error queue contains the substring
+// `substr_error`. Otherwise, returns `testing::AssertionFailure`.
+//
+// Defined as a predicate instead of a matcher since matchers must be purely
+// functional, but `ERR_get_error_line_data` has OpenSSL-internal side-effects.
+testing::AssertionResult IsOpenSslFailure(const char* return_expr,
+                                          const char* status_expr,
+                                          bool openssl_return,
+                                          const char* substr_error) {
+  auto expected = absl::StrFormat("false with OpenSSL error containing '%s'",
+                                  substr_error);
+  if (openssl_return) {
+    auto message = absl::StrFormat("%s unexpectedly returned true...",
+                                   return_expr);
+    return MakeFormattedAssertionFailure(message, expected, "true");
+  }
+
+  const char *file, *data;
+  int line, flags;
+  auto err = ERR_get_error_line_data(&file, &line, &data, &flags);
+
+  // If err == 0, then no error was available in the OpenSSL error queue.
+  if (!err) {
+    auto message = absl::StrFormat("%s returned false, but did not produce "
+                                   "OpenSSL error...", return_expr);
+    return MakeFormattedAssertionFailure(message, expected,
+                                         "false, but no OpenSSL error");
+  }
+
+  if (testing::Matches(testing::HasSubstr(substr_error))(data)) {
```

Comment:
Rather than hard-coding `HasSubstr` here, I'm wondering whether you can make the type of the `substr_error` argument a `testing::Matcher`. That way, the caller could specify any matcher on the string they wanted. e.g., they could choose to use `HasSubstr` or do exact matching.